### ISSUE TITLE
Active Task - Forms Tab

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#727]
Part 4 of Issue #404.
Prepare Task Forms tab contents.

![screen shot 2015-03-20 at 11 19 49 am](https://cloud.githubusercontent.com/assets/8228624/6758108/1070f734-cef3-11e4-8971-ec49ee448575.png)

![screen shot 2015-03-20 at 11 19 52 am](https://cloud.githubusercontent.com/assets/8228624/6758113/190ab614-cef3-11e4-8df6-7bc089ee58b3.png)

The two columns will be populated by forms used for the Task. Incomplete forms appear on the right. Completed forms will appear on the left side. We have populated the database with template and completed forms in order to create the lists.

The form elements are clickable. Clicking the form will display the form.
The breadcrumb in the header updates with the selected form.

**Comment by gsouvik** [comment]
@bsaiken 
I was planning to start with this issue next. After going through the available requirement details I have few questions:
1) I assume the right hand side data is populated from the table "task_form". The column "user_form_id" must be id of the connected user.
2) I could not find the table for "completed forms" to show data on the left hand column. Do I have to create the table, or I missed the db script?
3) I could not find the actual forms on local disk, which are populated in the "master_form" table.

It would be best if you could explain the requirement in details like you did for the task status tab issue. It was very descriptive and explained the requirement very clearly.

Thanks.

**Comment by bsaiken** [comment]
@gsouvik @FWAJL 

I apologize.  This wasn't completely set up for you.  We are still setting some things up that I thought would be ready by the time you got here.  I will add another comment to this Issue when we are ready.

The main thing I need is to identify where we are going to store Task Forms and Completed Task Forms.  I am waiting on Jeremie for the answer.  Jeremie and I may need to discuss this in our call Wednesday.

I am putting this Issue on hold, but the info below will help you understand our needs and you could even get started.  Fell free to move ahead to other Issues - I will reset priorities for you.

I am setting up an example under the username "Popup".  The task is:   "Example Task for Issue #727".

Please bear with me - we need to change some of our terminology to get this properly organized.  I will try to clearly explain as I go.

The right hand side will display Task Forms.

The relation table for Task Forms, `task_forms`, has not been create yet.  You will need to do that.  The table will have the following columns:
task_id
task-specific_form_id
location-specific_form_id 

FYI: I am having Jeremie rename the table `task_form` to `task_template_forms`.  This is because the table names (`task_form` and `task_forms`) are too similar and not sufficiently descriptive.  I need Jeremie to do this because the methods that populate this table also need to be modified (see Issue #787).  This does not affect your work.  

The two different kinds of Task Forms: Task-specific forms and Location-specific forms, shall be listed similar to the template forms (Master Forms and User Forms).    If you examine the Forms tab for the Task "Example Task for Issue #727" you will see the Master Forms and User Forms listed together, with "Master Forms" and "User Forms" shown as category headers.  So the right hand side will look like this, but with the category headers "Task-specific forms" and "Location-specific forms".  Because the output is similar, we may be able to use the same controller method to create the list - I will ask Jeremie about this.

**Comment by gsouvik** [comment]
Thanks @bsaiken for explaining in details. I shall return back to this task when more information is available. For now I shall pick another task from the list with highest priority and start with it.

**Comment by bsaiken** [comment]
I know where the forms shall be stored now.
All forms are stored in the /Web/uploads folder.
We have several different kind of forms, and it is useful to keep them in separate directories.  So, here is the list of form types and and their associated directories, and a brief explanation of each.

**Comment by gsouvik** [comment]
@bsaiken 
I was resuming with this particular issue today and had a minor question. The list of forms on this page should be presented in list box just like:
(After selecting a task and editing it)
http://localhost/FieldWorkAssistantMVC/task/forms
Please confirm.
Thanks.

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
I again tried started with this issue and ended up finding myself in some queries:

1)  So the right hand side under the heading "Forms used for this task" would fetch data from two tables - `document` and `task_template_form`. The records from `document` would be grouped under the sub heading "Location-specific forms" and those from `task_template_form` would be grouped as "Task-specific forms". Is my understanding till here correct?

2) Which table would be used for the left hand side or "Completed Forms".

3) What happens when the user selects a form from the right hand side and clicks on the center column button `from-categorized-list-right`. Do I have to copy the selected form into the directory "uploads/completed_forms". If yes what would be the naming convention in that case.

I might have some more queries as the requirement becomes from clear to me, but at present these few are the only ones.

Thank you.

**Comment by bsaiken** [comment]
@gsouvik 

1) We decided not to have "Task-specific forms" at this time.  So, there will only be Location-specific forms, and they don't need a category grouping - just display the list of forms.  The Location-specific forms are the ones you created in Issue #778.  The forms were generated from template, but they are now independent.  So the `task_template_form` table is no longer relevant.  You only need to list the forms where `document.document_id`==`task_location.location_id`.  
Although a Task can occur more than once, it will not be possible to have a Task active in two instances simultaneously, so the condition above must be unique.  
There may be more than one template that is copied for a given Task, so you may have more than one entry for the same `document.document_id`.  In this case, the `document_value` will be unique, so no duplicate entries are possible.  
The same Task could be, and often will be, done multiple times.  So, you may anticipate that we will have multiple copies of the completed forms - and we would.  In this case, we will need to add the date completed to the database table - but that is not part of the scope for this Issue.

2) The "Completed Forms" is not ready yet.  We need the mobile site developed so we can fill out the Location-specific forms.  Once the forms have been filled out, and submitted, we will have "completed" forms.  The form will be removed from the `/upload/task-location` directory to the `/upload/completed_forms` directory - that's how we know they have been completed. The contents of the `document` table will not change (except, as mentioned above, there will now be an entry for the date completed).   

3) There are not any buttons in this view.

Each form should have a contextual menu "view" link to review the form (but not fill it out).

**Comment by bsaiken** [comment]
@FWAJL 
Jeremie - Question.
If we have two Tasks that are occurring at the same time, and the Tasks share some of the same Task-locations, that could cause a problem, no?

**Comment by bsaiken** [comment]
@FWAJL 
I did not include you on the comment to Souvik - my mistake.  You need to go to github to see the comment to have context for my question about multiple simultaneous Tasks.

**Comment by FWAJL** [comment]
@gsouvik @bsaiken 

1) A correction:
> @bsaiken 
> You only need to list the forms where document.document_id==task_location.location_id.

It is document.document_category = "task_location_id" and use the document_value to find the task locations for a selected task (the id of the task locations is the first string in value before the "_").

2) We will need a completed_form table with a link for each task location to its document. Then when we reactivate a task on a recurring basis, we will search for the documents that not found in the completed_form table.
The document table is not an appropriate place to store completed form information.

Now, to @bsaiken : a completed form will always be only link to 1 task location, right? If so, we'll need the columns as follows: task_location_id, document_id, completed_form_status ("READY", "STARTED", "DONE").

Let me know
 
To answer @bsaiken last question:

If we have two tasks using the same locations, they will have the own task locations and so no problem.

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
Thank you both for the input. I've completed the coding as per the most updated comments. Please note I haven't added anything for the completed forms at this point as I understand that portion is not yet completed.

I have a question regarding the viewing of the form. Is this already done for any other issue already or I would have to develop this feature from the scratch. Also please share the details of the requirement like do I have to use any popup to show the forms etc.

**Comment by bsaiken** [comment]
@gsouvik 
I will answer your question first.  @mehedee1 already set this up using fancyapps.  The fancyapps js library is already uploaded to the site at /Web/js/addons/fancyapps.  I liked this solution just fine.  

Mehedee is working with me to pre-populate the pdf forms with information from the database, but he's not up to speed on our MVC code structure.  

The software he uses (pdftk) does not exist on Hurray Host, so he is not working through github.  

Still, he could give you some input on this Issue, if you need it.

**Comment by bsaiken** [comment]
@gsouvik 
I tried testing but I am seeing some problems.  Let me know if you are seeing the same things.
1) When I activate a task, I am having to refresh the page to see the activated task on the left hand side.  There should be a redirect so that the activate task list is automatically re-populated.
2) a m not seeing that the template files are being copied, and new files are created.
3) This may be to 2): The "Forms used for this task" list is empty.  The newly created forms should be listed.

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
We are accessing locations from Session. When session data was not available or not yet populated, the JS stalled. Fixed this issue by checking and populating Session data.

@mehedee1 Can you please share a link where I may review how the PDFs look when opened using fancyapps. Also please share any code you think, I might reuse.

Thank you.

**Comment by FWAJL** [comment]
@bsaiken @gsouvik 

Comment https://github.com/FWAJL/FieldWorkAssistantMVC/issues/727#issuecomment-94653011 updated as @bsaiken suggested.

**Comment by bsaiken** [comment]
@gsouvik @FWAJL 
I have tested the code on the feature_devsupport branch.  This is not working for me.  New forms are not generated.  I still have to refresh the page to see the task promoted to the active side.

When I try this on the live Hurray Host site, I do not need to refresh the page the new forms are generated.  However, however, the new forms do not appear on the active task forms tab.

Somehow, the live site is ahead of the feature branch.

Jeremie, I need the list to work before we update the Bluehost site, so we will need to update Bluehost again once this is done.

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
Please do another pull and test again. To be on the safer side I changed couple of lines of code.

On a separate note, it's very difficult to debug if I don't see the actual error message. Please post a screenshot of the AJAX call response from your browsers console so that I may know why the error occurs.

Also #778 is the main issue on which this issue depends and both should get fixed if this issue we are talking about is fixed.

Thank you.

**Comment by bsaiken** [comment]
@gsouvik 
I can't tell if there is a problem here, because activating Tasks requires a refresh, and does not create copes of the template files.  So maybe this would work, but without Issue #778 complete, I don't know.
Here are some screen shots to show how it is hanging up and requres a manual refresh.
I am bout to click "OK"...
![screen shot 2015-04-23 at 7 58 33 am](https://cloud.githubusercontent.com/assets/8228624/7300124/200a84b0-e98f-11e4-9a4a-ccc9c2135426.png)

Now I have clicked "OK" but the view is stuck:
![screen shot 2015-04-23 at 7 58 33 am](https://cloud.githubusercontent.com/assets/8228624/7300154/5ecc57b4-e98f-11e4-91dc-d173fbc28d60.png)

Refresh the page, and the task is promoted to an active task, but no form copes are made.
![screen shot 2015-04-23 at 7 58 48 am](https://cloud.githubusercontent.com/assets/8228624/7300171/7626c110-e98f-11e4-9417-41ad72fe542e.png)
![screen shot 2015-04-23 at 7 58 52 am](https://cloud.githubusercontent.com/assets/8228624/7300210/bd961168-e98f-11e4-8c9a-d32e1c720226.png)

**Comment by bsaiken** [comment]
@gsouvik 
When I check the live site, it works...

![screen shot 2015-04-23 at 8 11 16 am](https://cloud.githubusercontent.com/assets/8228624/7300386/ba69931a-e990-11e4-976a-2c9bce2ef595.png)

![screen shot 2015-04-23 at 8 13 06 am](https://cloud.githubusercontent.com/assets/8228624/7300407/dbf5191e-e990-11e4-87a9-7ac8771baa84.png)

The document table is populated...
![screen shot 2015-04-23 at 8 15 35 am](https://cloud.githubusercontent.com/assets/8228624/7300431/fca1cc66-e990-11e4-993d-7360e9dfd9ca.png)

The forms are created...
![screen shot 2015-04-23 at 8 16 18 am](https://cloud.githubusercontent.com/assets/8228624/7300443/0e603050-e991-11e4-9386-b99b818b8e25.png)
 
but the list of forms does not appear in the right hand column...
![screen shot 2015-04-23 at 8 17 00 am](https://cloud.githubusercontent.com/assets/8228624/7300464/32031748-e991-11e4-9f08-a564d09ee66b.png)
![screen shot 2015-04-23 at 8 18 31 am](https://cloud.githubusercontent.com/assets/8228624/7300497/59f2f264-e991-11e4-8ab6-757046917b40.png)

I am perplexed as to how if can work (at least partially) on the live site, but not on my localhost.

**Comment by gsouvik** [comment]
@bsaiken 
1) Please delete the Web/uploads/task_location directory and try it. If it still doesn't refresh, go to point 2. Also I assume you pulled from dev_support just before this test.

2) As soon as you click "OK" and it halts, network tab shows a call `updateitems`. Clicking on it Chrome should show you the details of the call and errors if any. Please post that screenshot here. Following is an example of how it should look:
![screen shot 2015-04-23 at 9 01 27 pm](https://cloud.githubusercontent.com/assets/2292482/7300791/f54849fc-e9fb-11e4-9c2d-e26624771dd3.png)

3) May we test this with other users too, it's strange why it works on one scenario but doesn't work on another system. Testing with others would help us pin point the issue.

**Comment by FWAJL** [comment]
@bsaiken @gsouvik 

I have tried Souvik's code and it works. When I activate a task with one template form and 4 locations, I see the forms in the Active UI.

Try the following:

1) Run the last database backup from Database/Scripts/Online backups/ (from April-21)
2) Download the whole upload folder from Hurrayhost as a zip, unzip it and replace your local folder.
3) Login as test PM
4) Select the EC-TX project and check the task Monitoring-2 so that it has the Well Purge template form.
5) Activate the task
6) Check the Form tab in the Active Task UI.

For me it worked. If you get the live database, you absolutely need the live upload folder with the documents in it. Otherwise, you'll get funny behaviors.

I hope this helps. LMK.

**Comment by bsaiken** [comment]
@FWAJL @gsouvik 
Here is where it hangs up.
![screen shot 2015-04-23 at 12 20 32 pm](https://cloud.githubusercontent.com/assets/8228624/7305514/355650a0-e9b3-11e4-81d1-48f3131df3f4.png)

I have having trouble testing other branches, too.  So it may be my machine.

**Comment by FWAJL** [comment]
@bsaiken @gsouvik 

Please click on updateItems and then response tab. We need to see the details of why it halted. 

Have you checked all the 6 points I mentioned above? Thank you.

**Comment by bsaiken** [comment]
@gsouvik @FWAJL 
See if this helps.
![screen shot 2015-04-23 at 2 04 50 pm](https://cloud.githubusercontent.com/assets/8228624/7307452/e1f8f232-e9c1-11e4-8983-d389db2afa15.png)

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
The only way that error is possible when the table `task_template_form` has data but NO matching data is available in `master_form`. The foreign key relation is not available and the script is not finding any matching master form corresponding to the task template.

As @FWAJL suggested in earlier messages, please start from the beginning, creating master forms, then assigning them to tasks and then promoting the task to active. Let us know if that helped.

Thank you.

**Comment by FWAJL** [comment]
@bsaiken @gsouvik 

Just FYI: the copy of the forms at task activation only works when we copy master form. Copying user forms doesn't work because it is not implemented.

So @bsaiken : are you using a user form in your case? Thank you.

**Comment by bsaiken** [comment]
@fwajl @gsouvik

Jeremie, I believe I tried everything you suggested. Since you are certain it works, in this case skip my review and I will review when it goes live on the server. 

We do need user form implementation.  The only difference between a user form and a master form is who uploads the form: master (admin); user (PM). 

Don't get too generic about copying forms. We will eventually have to deal with the case where the template is only copied once. For example, the chain of custody, which will be a commonly used form, will not get copied once for each location. Instead, all the location names will be put on the form. I do NOT want to implement this now, but it is important to know.

**Comment by FWAJL** [comment]
@bsaiken @gsouvik 

New issue is required I think what is described above > issue #906

**Comment by bsaiken** [comment]
@gsouvik 
Can you complete this so that the generated forms show in the list.  That's top priority.

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
But this is already tested to be working on my and @FWAJL computer. This could not be reproduced on your end because of issue #906 . That's what I meant in my comment in that thread. If we fix #906, implicitly it would fix this and #778.

As an issue there is there is nothing more to code in this. Let me know your thoughts.

**Comment by bsaiken** [comment]
@FWAJL @gsouvik 
Jeremie,  I am OK viewing this on the live site, so you could release the code.

**Comment by WebDevJL** [comment]
@bsaiken 

To do 100% sure, you are ok to release the code where if a task has a form that is referencing a user form, it will not activate. If the task form are just master form, it will work.
![task-with-user-form](https://cloud.githubusercontent.com/assets/3506074/7363903/7b0e176a-ed80-11e4-96e3-7e3eb6d4e400.png)
![task-activation-with-user-form-fails](https://cloud.githubusercontent.com/assets/3506074/7363909/7d5cf7fc-ed80-11e4-8feb-4e486843d04d.png)

If you are positive, I will release tomorrow. Thank you.

**Comment by bsaiken** [comment]
@FWAJL 
Absolutely release it.  This Issue is holding up others.  We cannot afford that right now.
All that is needed is that only master forms get used.  Since, in the beginning, I will be the only user, it is a problem we can fix later.

**Comment by bsaiken** [comment]
@FWAJL 
Jeremie,
I just tried testing this again, and it does not work on my local machine.  So, if it's working for you, I definitely need to see this on the live site.  I rearranged your priorities so this is the highest priority.

**Comment by WebDevJL** [comment]
@bsaiken 

The current of items to be released and this one will be released tomorrow morning. 

Thank you.

**Comment by bsaiken** [comment]
Super - thanks.

Brian
